### ISSUE TITLE
[feature] User-level domain blocks

### ISSUE THREAD

**Issue by VyrCossont** [superseriousbusiness/gotosocial#3061]
### Is your feature request related to a problem ?

As a non-admin user, I want to not have any interactions with a given instance, see posts or accounts from it, or have my posts sent there.

### Describe the solution you'd like.

I'd like to see GtS implement Mastodon-compatible user-level domain blocks. Until very recently, Mastodon implemented them as more of a domain mute, but as of 2022, Mastodon instances with authorized fetch turned on will not serve a user's posts to domains that they have blocked. Since we always require authorization, we would do the same.

- [Mastodon docs for this feature](https://docs.joinmastodon.org/user/moderating/#block-domain)
- [Claire's comment on how it interacts with authorized fetch](https://github.com/mastodon/mastodon/issues/18345#issuecomment-1120176872)
- [Mastodon client API](https://docs.joinmastodon.org/methods/domain_blocks/)

Domain blocks do not federate but have side effects that do.

A user blocking a domain should:
- sever any existing follow relationships between the blocking user and accounts on that domain, in either direction, which generates `Reject` activities
- reject any pending follow requests for the blocking user from accounts on that domain, which also generates `Reject` activities
- remove any notifications for the blocking user from accounts on that domain
- leave single-user blocks and mutes alone (in case the domain-blocking user later changes their mind and unblocks the blocked domain, we want to leave any more specific blocks in place)
- set the `domain_blocking` flag on `Relationship` API objects related to accounts on the blocked domain
- hide accounts from that domain as if each account was blocked (serving a blocked-account stub if we would do that for a normal block)
- hide statuses from that domain in timelines, search results, favs, bookmarks, etc. as if the author of each status was blocked
- at time of block, invalidate account and status visibility caches for the blocking user
- prevent the blocking user's instance from serving the user's actor or activities to the blocked domain, except for the `Reject` activities mentioned above

Blocking a domain should also apply to any subdomains of that domain, the same way that our admin-level domain blocks work. I don't believe this is the case for Mastodon's own implementation, but I want to improve on that.

Documentation should reflect the same caveats as [our admin-level domain blocks docs](https://docs.gotosocial.org/en/latest/admin/domain_blocks/): users on blocked domains can still read your public statuses on the web frontend, RSS (if enabled), other instances, etc.

### Describe alternatives you've considered.

For very small instances, admin domain blocks are equivalent. This feature would be useful for instances where there are enough users that they don't all agree on which domains they don't want to interact with: some users may want to block instances such as `mastodon.social` that other users may grudgingly tolerate.

### Additional context.

This is lower priority than our current roadmap issues, but interacts tangentially with relay support, because subscribing to a relay may pull a lot more posts from annoying domains into the instance's federated timeline.

**Comment by daenney** [comment]
We've migrated to Codeberg. Your issue has been imported as https://codeberg.org/superseriousbusiness/gotosocial/issues/3061.
